qiskit/transpiler/passes/lookahead_mapper.py

```diff
@@ -0,0 +1,282 @@
+# -*- coding: utf-8 -*-
+
+# Copyright 2018, IBM.
+#
+# This source code is licensed under the Apache License, Version 2.0 found in
+# the LICENSE.txt file in the root directory of this source tree.
+
+"""
+Implementation of Sven Jandura's swap mapper submission for the 2018 QISKit
+Developer Challenge, adapted to integrate into the transpiler architecture.
+
+The role of the mapper pass is to modify the starting circuit to be compatible
+with the target device's topology (the set of two-qubit gates available on the
+hardware.) To do this, the mapper will insert SWAP gates to relocate the logical
+qubits for each upcoming gate onto a set of coupled physical qubits. However, as
+SWAP gates are particularly lossy, the goal is to accomplish this remapping while
+introducing the fewest possible additional SWAPs.
+
+This algorithm searches through the available combinations of SWAP gates by means
+of a narrowed best first/beam search, described as follows:
+
+- Start with a layout of logical qubits onto physical qubits.
+- Find any gates in the input circuit which can be performed with the current
+  layout and mark them as mapped.
+- For all possible SWAP gates, calculate the layout that would result from their
+  application and rank them according to the distance of the resulting layout
+  over upcoming gates (see _calc_layout_distance.)
+- For the four (SEARCH_WIDTH) highest-ranking SWAPs, repeat the above process on
+  the layout that would be generated if they were applied.
+- Repeat this process down to a depth of four (SEARCH_DEPTH) SWAPs away from the
+  initial layout, for a total of 256 (SEARCH_WIDTH^SEARCH_DEPTH) prospective
+  layouts.
+- Choose the layout which maximizes the number of two-qubit which could be
+  performed. Add its mapped gates, including the SWAPs generated, to the
+  output circuit.
+- Repeat the above until all gates from the initial circuit are mapped.
+
+For more details on the algorithm, see Sven's blog post:
+https://medium.com/qiskit/improving-a-quantum-compiler-48410d7a7084
+
+"""
+
+from qiskit.dagcircuit import DAGCircuit
+from qiskit.transpiler._basepasses import TransformationPass
+from qiskit.mapper import remove_last_measurements, return_last_measurements
+
+SEARCH_DEPTH = 4
+SEARCH_WIDTH = 4
+
+
+class LookaheadMapper(TransformationPass):
+    """Map input circuit onto a backend topology via insertion of SWAPs."""
+
+    def __init__(self, coupling_map):
```

Comment:
can we call this coupling_graph @ajavadia what do you think?

Reply:
No stance on the change, but referencing #1299 so it's included across mappers.